src/Data/Text/Builder/Linear/Dec.hs

```diff
@@ -170,3 +182,99 @@ quotBillion a = case (finiteBitSize a, isSigned a) of
     cast f = fromIntegral (f (fromIntegral a))
 #endif
 {-# INLINEABLE quotBillion #-}
+
+--------------------------------------------------------------------------------
+-- Unbounded
+--------------------------------------------------------------------------------
+
+-- | Append /unbounded/ decimal number.
+(|>$$) ∷ Integral a ⇒ Buffer ⊸ a → Buffer
+
+infixl 6 |>$$
+buffer |>$$ n = case toInteger n of
+  n' →
+    appendBounded
+      (maxIntegerDecLen n')
+      (\dst dstOff → unsafeAppendUnboundedDec dst dstOff n')
+      buffer
+{-# INLINEABLE (|>$$) #-}
+
+-- | Prepend /unbounded/ decimal number.
+($$<|) ∷ Integral a ⇒ a → Buffer ⊸ Buffer
+
+infixr 6 $$<|
+n $$<| buffer = case toInteger n of
+  n' →
+    prependBounded
+      (maxIntegerDecLen n')
+      (\dst dstOff → unsafePrependUnboundedDec dst dstOff n')
+      (\dst dstOff → unsafeAppendUnboundedDec dst dstOff n')
+      buffer
+{-# INLINEABLE ($$<|) #-}
+
+-- | ceiling (fbs a * logBase 10 2) < ceiling (fbs a * 5 / 16) < 1 + floor (fbs a * 5 / 16)
+maxIntegerDecLen ∷ Integer → Int
+maxIntegerDecLen a = case a of
+  I.IS i# → maxDecLen (I# i#)
+  I.IP n# → 1 + (fromIntegral (BN.bigNatSize n#) * finiteBitSize (0 ∷ Word) * 5) `shiftR` 4
+  I.IN n# → 2 + (fromIntegral (BN.bigNatSize n#) * finiteBitSize (0 ∷ Word) * 5) `shiftR` 4
+{-# INLINEABLE maxIntegerDecLen #-}
+
+exactIntegerDecLen ∷ Integer → Int
+exactIntegerDecLen n = case n of
+  I.IS i# → exactDecLen (I# i#)
+  I.IP n# → fromIntegral (BN.bigNatSizeInBase 10 n#)
+  I.IN n# → 1 + fromIntegral (BN.bigNatSizeInBase 10 n#)
+
+unsafeAppendUnboundedDec ∷ A.MArray s → Int → Integer → ST s Int
+unsafeAppendUnboundedDec marr off n = unsafePrependUnboundedDec marr (off + exactIntegerDecLen n) n
+{-# INLINEABLE unsafeAppendUnboundedDec #-}
+
+unsafePrependUnboundedDec ∷ ∀ s. A.MArray s → Int → Integer → ST s Int
+unsafePrependUnboundedDec marr !off n = case toInteger n of
+  I.IS i# → unsafePrependBoundedDec marr off (I# i#)
+  I.IP n# → go (off - 1) n# >>= \o → pure (off - o)
+  I.IN n# → go (off - 1) n# >>= sign
+  where
+    sign !o = do
+      A.unsafeWrite marr (o - 1) 0x2d -- '-'
+      pure (off - o + 1)
+
+    -- TODO: remove once best implementation found
+    -- go ∷ Int → BN.BigNat# → ST s Int
+    -- go o n# = case BN.bigNatCompareWord# n# 10## of
+    --   LT → do
+    --     A.unsafeWrite marr o (fromIntegral (0x30 + BN.bigNatToWord n#))
+    --     pure o
+    --   _ → do
+    --     let !(# q#, r## #) = BN.bigNatQuotRemWord# n# 100##
+    --     A.copyFromPointer marr (o - 1) (digits `plusPtr` (fromIntegral (W# r##) `shiftL` 1)) 2
+    --     case BN.bigNatCompareWord# n# 100## of
+    --       LT → pure (o - 1)
+    --       _ → go (o - 2) q#
+
+    go ∷ Int → BN.BigNat# → ST s Int
+    go o n# = do
+      let !(# q#, r## #) = BN.bigNatQuotRemWord# n# 100##
```

Comment:
This is quadratic: if `n` is `m` bits long division is O(m) and repeats in a loop O(m) times again. Divide-and-conquer approach from [`fast-digits`](https://hackage.haskell.org/package/fast-digits-0.3.2.0/docs/Data-FastDigits.html) should work better.